node/primitives/src/lib.rs

```diff
@@ -56,6 +57,112 @@ impl Statement {
 	}
 }
 
+/// This helper trait ensures that we can encode Statement as CompactStatement,
+/// and anything as itself.
+pub trait EncodeAs<T> {
+	fn encode_as(&self) -> Vec<u8>;
+}
+
+impl<T: Encode> EncodeAs<T> for T {
+	fn encode_as(&self) -> Vec<u8> {
+		self.encode()
+	}
+}
+
+/// A statement about the validity of a parachain candidate.
+///
+/// This variant should only be used in the production of `SignedStatement`s. The only difference between
+/// this enum and `Statement` is that the `Seconded` variant contains a `Hash` instead of a `CandidateReceipt`.
+/// The rationale behind the difference is that a `CandidateReceipt` contains `HeadData`, which does not have
+/// bounded size. By using this enum instead, we ensure that the production and validation of signatures is fast
+/// while retaining their necessary cryptographic properties.
+#[derive(Debug, Clone, PartialEq, Encode, Decode)]
+enum CompactStatement {
+	/// A statement about a new candidate being seconded by a validator. This is an implicit validity vote.
+	#[codec(index = "1")]
+	Seconded(Hash),
+	/// A statement about the validity of a candidate, based on candidate's hash.
+	#[codec(index = "2")]
+	Valid(Hash),
+	/// A statement about the invalidity of a candidate.
+	#[codec(index = "3")]
+	Invalid(Hash),
+}
+
+impl EncodeAs<CompactStatement> for Statement {
+	fn encode_as(&self) -> Vec<u8> {
+		let statement = match *self {
+			Statement::Seconded(ref c) => {
+				polkadot_primitives::parachain::Statement::Candidate(c.hash())
+			}
+			Statement::Valid(hash) => polkadot_primitives::parachain::Statement::Valid(hash),
+			Statement::Invalid(hash) => polkadot_primitives::parachain::Statement::Invalid(hash),
+		};
+		statement.encode()
+	}
+}
+
+/// A signed type which encapsulates the common desire to sign some data and validate a signature.
+///
+/// Note that the internal fields are not public; they are all accessable by immutable getters.
+/// This reduces the chance that they are accidentally mutated, invalidating the signature.
+pub struct Signed<Payload, RealPayload = Payload> {
```

Comment:
Should this be accompanied by a  `pub type SignedStatement` and `pub type SignedAvailabilityBitfield`?

Should probably go in the root `primitives` crate as runtime will want access to those types - it won't have access to this crate.

Reply:
Yeah, I already moved the implementation; I noticed the same thing. I also did add a `pub type SignedStatement`; I'm going to do the `SignedAvailabilityBitfield` next.